src/Bladeburner/Actions/TeamCasualties.ts

```diff
@@ -0,0 +1,49 @@
+import type { getRandomIntInclusive } from "../../utils/helpers/getRandomIntInclusive";
+
+export enum CasualtyFactor {
+  LOW_CASUALTIES = 0.5, // 50%
+  HIGH_CASUALTIES = 1, // 100%
+}
+
+export interface OperationTeam {
+  teamSize: number;
+  teamLost: number;
+
+  killSupportingSleeves(sleeveDeaths: number): void;
+}
+
+/**
+ * Some actions (Operations and Black Operations) use teams for success bonus
+ * and may result in casualties, reducing the player's hp, killing team members
+ * and killing sleeves (to shock them, sleeves are immortal)
+ *
+ * Caveats: TeamSize = Human Team + Supporting Sleeves set to this action
+ */
+export class TeamCasualties {
+  constructor(private severity: CasualtyFactor, private teamSizeUsed: number, private supportingSleeves: number) {}
+
+  rollOutcome(random: typeof getRandomIntInclusive, team: OperationTeam) {
```

Comment:
I don't like this 'throwaway' class. If you'd like to parametrize, and then keep it this way, you can do so using a function returning a parametrized function. But you're using it as follows:

```
const casualties = new TeamCasualties(severity, action.teamCount, this.sleeveSize);
return casualties.rollOutcome(getRandomIntInclusive, this);
```
casualities get's then thrown away the next time the garbage collector runs. The only thing that I can imagine that would somehow make sense to keep parametrized is random, the random number generator, to be able too keep the events deterministic. But that's currently not a thing in the whole codebase. Also, you are using another random function to shuffle the array, so having one as a parameter seems not consistent. What was your intent?

I'd recommend to drop the random parameter, and just use the stack for all 4 remaining arguments: Make this function static. 

What I really like is the interface for the BladeBurner object (OperationTeam)!

Reply:
I agree. This is a case of making the implementation awkward in order to try and do testing.

Note I say "try" because an important precept of unit testing is that you should be testing the *public* interfaces of units, not internal implementation details. This is because the internal details can change, via refactoring; trying to test the internal details is too granular, and also doesn't cover what is actually important, which is the public API. Sometimes you need to expose internal knobs in order to make mocks and make testing easier or possible, but that is different from testing the internals directly.

Reply:
Regarding the static—I feel like I'm playing devil's advocate here but there isn't a clean solution for this. We have 2 components merging into data flow: the rolls (low, high, size, used), and the outcomes (randomGenerator, deaths, sleeves killed (specifically)).

If I make it static, then the function call has to be curried for it to have segmented return values. If I don't curry it, then the final return injects test-only data into the production code. The class was a sort of middle ground as d0s already commented on this, but I'm afraid I can't make both of you happy.

I disagree on the garbage collection aspect, this calculation is no different than using objects for actionId or returning [nextState, prevState] object arrays in corporations. The _state_ object is a byproduct of the actionObject, which gets created once per action, with relevant bits cloned. This is one of those "cloned" bits.

The separation of the interfaces is specific to allow the random generator to be injected without mocking at test time as the randomness of things in Bladeburner are notoriously error-prone.

Reply:
I agree with d0sboots about the mocking of Math.random. In its current state, the custom random function is only useful for testing while it reduces the readability. I don't think it's worth the trade-off.

Reply:
I don't follow. I'm replacing the outside interface here, not the internal one. _Random_ is the internal bit. I didn't want to expose _how_ Math.random (float) gets converted to a range. Which is why I allowed the higher-cohesion unit to be injected and stubbed (getRandomIntInclusive) rather than Math.random.

On the topic of testing the *public* interfaces, I disagree. The codebase is not in a state where each publicly accessible component can be isolated and controlled. We'll get there, but it may take a few hundred PRs. For the time being I'm increasing what little coverage we can get in lean pieces of code **without explicitly touching** the bladeburner public API, so that these two worlds won't mix.

I find it difficult to even talk about public/internal/external APIs when for most classes all accessors are public and there's little notion of where the boundaries are (case in point: Player, Sleeves, Bladeburner).

But anyway—what are we doing about this? I can give it another design round to try to make everyone happy.

Reply:
Okay, I see the conversation diverged a bit. There's 4 issues:
- the surface area of the ROLLING random generator
- the surface area of the SHUFFLING SLEEVES
- the surface area of the Casualties constructor (and calls)
- whether to keep this single-function behavior in a class or function (static or no)

Which one here is most important to get this PR moving forward?

Reply:
> I don't follow. I'm replacing the outside interface here, not the internal one. _Random_ is the internal bit. I didn't want to expose _how_ Math.random (float) gets converted to a range. Which is why I allowed the higher-cohesion unit to be injected and stubbed (getRandomIntInclusive) rather than Math.random.

Why not use and mock getRandomIntInclusive then?
Same for the shuffle function.

Reply:
> If I make it static, then the function call has to be curried for it to have segmented return values. If I don't curry it, then the final return injects test-only data into the production code. The class was a sort of middle ground as d0s already commented on this, but I'm afraid I can't make both of you happy.

Here I cannot really follow. I would't curry here at all, have a simple function that returns deaths and a list / a count of all damaged sleeves. To test the function, just create an object that implements the Operation team interface , and mock random/ getRandomIntInclusive / shuffle.

Reply:
What I mean is giving it the following signature:
```ts
static TeamCasualtiesRollOutcome(team: OperationTeam, severity: CasualtyFactor, teamSizeUsed: number, supportingSleeves: List<SleeveObject>) {
... 
return {deaths, damagedSleeves } 
}
```
Or, to simplify things further, put teamSizeUsed and supportingSleeves inside the OperationTeam interface.
Maybe even make it a class, put the roll function in there, to stay OO.

Reply:
That path already got rejected. And we did the mock version as well in the first two comits, the mock got rejected as well.

Look I appreciate what you guys are doing but it gets tricky when all of a sudden everyone passionately shows up and wants to chip in with sentimental nits or personal preferences.

I presented a working solution with a very small compromise. Non of your suggestions are compromise-free.

Ie. if I follow your suggestion, then the method goes from 3 parameters to 4, and to 5 if I hook in the random or the mock. If I replace that with an object because the method has too many parameters I end up here again.

I'll zoom out and consider an alternative approach.

Reply:
> That path already got rejected. And we did the mock version as well in the first two comits, the mock got rejected as well.

Sorry if that is the case, but I cannot find the rejection. Can you provide a link to that? I somehow lost the overview here, too many code reviews. I also struggle to read them, since somehow the code lines are messed up. I could imagine the force pushes overwrote the reviewed commit on the server.

Reply:
> Look I appreciate what you guys are doing but it gets tricky when all of a sudden everyone passionately shows up and wants to chip in with sentimental nits or personal preferences.

I've been trying to keep my comments down for this reason. XD

There's some stuff I like, and some stuff I don't, but I don't have a compromise-free solution to suggest either. I'll try to break some more time free later today and see if I can come up with a more positive proposal, as opposed to "I don't like some of this." :D